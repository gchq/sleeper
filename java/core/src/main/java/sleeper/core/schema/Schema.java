/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.schema;

import sleeper.core.schema.type.PrimitiveType;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.TreeSet;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * A Schema describes the fields found in a particular table in a Sleeper instance.
 */
public class Schema {
    private final List<Field> rowKeyFields;
    private final List<Field> sortKeyFields;
    private final List<Field> valueFields;

    private Schema(Builder builder) {
        rowKeyFields = validateRowKeys(builder.rowKeyFields);
        sortKeyFields = validateSortKeys(builder.sortKeyFields);
        valueFields = validateValueKeys(builder.valueFields);
        validateNoDuplicates(streamAllFields(rowKeyFields, sortKeyFields, valueFields));
    }

    public static Builder builder() {
        return new Builder();
    }

    /**
     * Creates an instance of this class using a schema file.
     *
     * @param  schemaPath  the path to the schema file
     * @return             an instance of this class
     * @throws IOException if an error occurs reading from the file
     */
    public static Schema load(Path schemaPath) throws IOException {
        return loadFromString(Files.readString(schemaPath));
    }

    /**
     * Creates an instance of this class using a JSON string.
     *
     * @param  schemaJson the JSON string representing a schema
     * @return            an instance of this class
     */
    public static Schema loadFromString(String schemaJson) {
        return new SchemaSerDe().fromJson(schemaJson);
    }

    public List<Field> getRowKeyFields() {
        return rowKeyFields;
    }

    public List<Field> getSortKeyFields() {
        return sortKeyFields;
    }

    public List<Field> getValueFields() {
        return valueFields;
    }

    public List<PrimitiveType> getRowKeyTypes() {
        return getMappedFields(rowKeyFields, f -> (PrimitiveType) f.getType());
    }

    public List<PrimitiveType> getSortKeyTypes() {
        return getMappedFields(sortKeyFields, f -> (PrimitiveType) f.getType());
    }

    public List<String> getRowKeyFieldNames() {
        return getMappedFields(rowKeyFields, Field::getName);
    }

    public List<String> getSortKeyFieldNames() {
        return getMappedFields(sortKeyFields, Field::getName);
    }

    public List<String> getValueFieldNames() {
        return getMappedFields(valueFields, Field::getName);
    }

    public List<String> getAllFieldNames() {
        return getMappedFields(streamAllFields(), Field::getName);
    }

    private <T> List<T> getMappedFields(List<Field> fields, Function<Field, T> mapping) {
        return getMappedFields(fields.stream(), mapping);
    }

    private <T> List<T> getMappedFields(Stream<Field> fields, Function<Field, T> mapping) {
        return fields
                .map(mapping)
                .collect(Collectors.toUnmodifiableList());
    }

    public List<Field> getAllFields() {
        return streamAllFields().collect(Collectors.toUnmodifiableList());
    }

    /**
     * Streams all fields in the schema. This includes row key, sort key, and value fields.
     *
     * @return a stream of all fields
     */
    public Stream<Field> streamAllFields() {
        return streamAllFields(rowKeyFields, sortKeyFields, valueFields);
    }

    /**
     * Gets a field by name.
     *
     * @param  fieldName the name of the field
     * @return           a field, or an empty optional if the field does not exist
     */
    public Optional<Field> getField(String fieldName) {
        return streamAllFields()
                .filter(f -> f.getName().equals(fieldName))
                .findFirst();
    }

    /**
     * Saves this schema to a file.
     *
     * @param  path        the path of the file
     * @throws IOException if an error occured when writing the file
     */
    public void save(Path path) throws IOException {
        Files.writeString(path, new SchemaSerDe().toJson(this));
    }

    @Override
    public String toString() {
        return "Schema{" + "rowKeyFields=" + rowKeyFields + ", sortKeyFields=" + sortKeyFields + ", valueFields=" + valueFields + '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Schema schema = (Schema) o;

        return Objects.equals(rowKeyFields, schema.rowKeyFields) &&
                Objects.equals(sortKeyFields, schema.sortKeyFields) &&
                Objects.equals(valueFields, schema.valueFields);
    }

    @Override
    public int hashCode() {
        return Objects.hash(rowKeyFields, sortKeyFields, valueFields);
    }

    /**
     * Builder to create a schema object.
     */
    public static final class Builder {
        private List<Field> rowKeyFields;
        private List<Field> sortKeyFields;
        private List<Field> valueFields;

        private Builder() {
        }

        /**
         * Sets the row key fields.
         *
         * @param  rowKeyFields the row key fields
         * @return              the builder
         */
        public Builder rowKeyFields(List<Field> rowKeyFields) {
            this.rowKeyFields = rowKeyFields;
            return this;
        }

        /**
         * Sets the row key fields.
         *
         * @param  rowKeyFields the row key fields
         * @return              the builder
         */
        public Builder rowKeyFields(Field... rowKeyFields) {
            return rowKeyFields(Arrays.asList(rowKeyFields));
        }

        /**
         * Sets the sort key fields.
         *
         * @param  sortKeyFields the sort key fields
         * @return               the builder
         */
        public Builder sortKeyFields(List<Field> sortKeyFields) {
            this.sortKeyFields = sortKeyFields;
            return this;
        }

        /**
         * Sets the sort key fields.
         *
         * @param  sortKeyFields the sort key fields
         * @return               the builder
         */
        public Builder sortKeyFields(Field... sortKeyFields) {
            return sortKeyFields(Arrays.asList(sortKeyFields));
        }

        /**
         * Sets the value fields.
         *
         * @param  valueFields the value fields
         * @return             the builder
         */
        public Builder valueFields(List<Field> valueFields) {
            this.valueFields = valueFields;
            return this;
        }

        /**
         * Sets the value fields.
         *
         * @param  valueFields the value fields
         * @return             the builder
         */
        public Builder valueFields(Field... valueFields) {
            return valueFields(Arrays.asList(valueFields));
        }

        public Schema build() {
            return new Schema(this);
        }
    }

    private static List<Field> validateRowKeys(List<Field> fields) {
        if (fields == null || fields.isEmpty()) {
            throw new IllegalArgumentException("Must have at least one row key field");
        }
        if (fields.stream().anyMatch(field -> !(field.getType() instanceof PrimitiveType))) {
            throw new IllegalArgumentException("Row key fields must have a primitive type");
        }
        return makeImmutable(fields);
    }

    private static List<Field> validateSortKeys(List<Field> fields) {
        if (fields == null) {
            return Collections.emptyList();
        }
        if (fields.stream().anyMatch(field -> !(field.getType() instanceof PrimitiveType))) {
            throw new IllegalArgumentException("Sort key fields must have a primitive type");
        }
        return makeImmutable(fields);
    }

    private static List<Field> validateValueKeys(List<Field> fields) {
        if (fields == null) {
            return Collections.emptyList();
        }
        return makeImmutable(fields);
    }

    private static List<Field> makeImmutable(List<Field> fields) {
        return Collections.unmodifiableList(new ArrayList<>(fields));
    }

    private static Stream<Field> streamAllFields(
            List<Field> rowKeyFields, List<Field> sortKeyFields, List<Field> valueFields) {
        return Stream.of(rowKeyFields, sortKeyFields, valueFields)
                .flatMap(List::stream);
    }

    private static void validateNoDuplicates(Stream<Field> fields) {
        Set<String> foundNames = new HashSet<>();
        Set<String> duplicates = new TreeSet<>();
        fields.forEach(field -> {
            boolean isNew = foundNames.add(field.getName());
            if (!isNew) {
                duplicates.add(field.getName());
            }
        });
        if (!duplicates.isEmpty()) {
            throw new IllegalArgumentException("Found duplicate field names: " + duplicates);
        }
    }
}
