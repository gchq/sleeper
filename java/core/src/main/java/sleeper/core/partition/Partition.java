/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import sleeper.core.key.Key;
import sleeper.core.range.Region;
import sleeper.core.range.RegionCanonicaliser;
import sleeper.core.schema.Schema;

import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

/**
 * A Partition is a Region of key space, with additional information that allows
 * the position of the partition in the tree of partitions to be identified,
 * e.g. it has a unique id, the id of its parent partition, the id of any child
 * partitions, and if it has been split the dimension it was split on.
 * <p>
 * The {@link Region} must be in canonical form, i.e. all the ranges must contain
 * their minimum but not contain their maximum.
 */
public class Partition {
    private final Region region;
    private final String id;
    private final boolean leafPartition;
    private final String parentPartitionId;
    private final List<String> childPartitionIds;
    private final int dimension;

    private Partition(Partition.Builder builder) {
        region = Objects.requireNonNull(builder.region, "region must not be null");
        id = Objects.requireNonNull(builder.id, "id must not be null");
        leafPartition = builder.leafPartition;
        parentPartitionId = builder.parentPartitionId;
        childPartitionIds = Optional.ofNullable(builder.childPartitionIds).orElse(Collections.emptyList());
        dimension = builder.dimension;
        if (!RegionCanonicaliser.isRegionInCanonicalForm(region)) {
            throw new IllegalArgumentException("Region must be in canonical form");
        }
    }

    public static Builder builder() {
        return new Builder();
    }

    /**
     * Checks if a row key is in this partition. This depends on the ranges the partition covers.
     *
     * @param  schema schema of the Sleeper table
     * @param  rowKey values of row key to check
     * @return        true if the row key is in the partition
     */
    public boolean isRowKeyInPartition(Schema schema, Key rowKey) {
        return region.isKeyInRegion(schema, rowKey);
    }

    /**
     * Checks if a region overlaps this partition.
     *
     * @param  otherRegion region to check
     * @return             true if any part of the region is in this partition
     */
    public boolean doesRegionOverlapPartition(Region otherRegion) {
        return region.doesRegionOverlap(otherRegion);
    }

    public Region getRegion() {
        return region;
    }

    public String getId() {
        return id;
    }

    public boolean isLeafPartition() {
        return leafPartition;
    }

    public String getParentPartitionId() {
        return parentPartitionId;
    }

    public List<String> getChildPartitionIds() {
        return childPartitionIds;
    }

    public int getDimension() {
        return dimension;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Partition partition = (Partition) o;

        return leafPartition == partition.leafPartition
                && Objects.equals(region, partition.region)
                && Objects.equals(id, partition.id)
                && Objects.equals(parentPartitionId, partition.parentPartitionId)
                && Objects.equals(childPartitionIds, partition.childPartitionIds)
                && dimension == partition.getDimension();
    }

    @Override
    public int hashCode() {
        return Objects.hash(region, id, leafPartition,
                parentPartitionId, childPartitionIds, dimension);
    }

    @Override
    public String toString() {
        return "Partition{"
                + "region=" + region
                + ", id='" + id + '\''
                + ", leafPartition=" + leafPartition
                + ", parentPartitionId='" + parentPartitionId + '\''
                + ", childPartitionIds=" + childPartitionIds
                + ", dimension=" + dimension
                + '}';
    }

    public Builder toBuilder() {
        return builder()
                .region(region)
                .id(id)
                .leafPartition(leafPartition)
                .childPartitionIds(childPartitionIds)
                .parentPartitionId(parentPartitionId)
                .dimension(dimension);
    }

    /**
     * Builder to create a partition object.
     */
    public static final class Builder {
        private Region region;
        private String id;
        private boolean leafPartition;
        private String parentPartitionId;
        private List<String> childPartitionIds;
        private int dimension = -1; // -1 used to indicate that it has not been split yet; when it has been split, indicates which dimension was used to split on.

        private Builder() {
        }

        /**
         * Sets the region covered by the partition.
         *
         * @param  region the region
         * @return        the builder
         */
        public Builder region(Region region) {
            this.region = region;
            return this;
        }

        /**
         * Sets the ID of the partition.
         *
         * @param  id a unique identifier
         * @return    the builder
         */
        public Builder id(String id) {
            this.id = id;
            return this;
        }

        /**
         * Sets whether this is a leaf partition.
         *
         * @param  leafPartition true if the partition has no child partitions, false otherwise
         * @return               the builder
         */
        public Builder leafPartition(boolean leafPartition) {
            this.leafPartition = leafPartition;
            return this;
        }

        /**
         * Sets the parent of the partition. This is the ID of the partition that was split to create this partition.
         * Can be null if this is the root partition of the Sleeper table.
         *
         * @param  parentPartitionId the ID of the parent partition
         * @return                   the builder
         */
        public Builder parentPartitionId(String parentPartitionId) {
            this.parentPartitionId = parentPartitionId;
            return this;
        }

        /**
         * Sets the children of the partition. This is the IDs of any partitions that were created by splitting this
         * partition. This will default to an empty list, and must be empty for a leaf partition.
         *
         * @param  childPartitionIds the IDs of the child partitions
         * @return                   the builder
         */
        public Builder childPartitionIds(List<String> childPartitionIds) {
            this.childPartitionIds = childPartitionIds;
            return this;
        }

        /**
         * Sets the dimension this partition was split on. If this partition has been split, this should be the index of
         * the row key in the schema that this partition was split on.
         *
         * @param  dimension the index of the row key used as the split point
         * @return           the builder
         */
        public Builder dimension(int dimension) {
            this.dimension = dimension;
            return this;
        }

        public Partition build() {
            return new Partition(this);
        }

        public String getId() {
            return id;
        }

        public Region getRegion() {
            return region;
        }
    }
}
