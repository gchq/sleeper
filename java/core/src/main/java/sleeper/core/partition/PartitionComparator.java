/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import sleeper.core.schema.Schema;

import java.util.Arrays;
import java.util.Comparator;

/**
 * Comparator for organised paritions in correct orientation.
 */
public class PartitionComparator implements Comparator<Partition> {

    private Schema schema;

    PartitionComparator(Schema schema) {
        this.schema = schema;
    }

    @Override
    public int compare(Partition partition1, Partition partition2) {
        String rowKey = schema.getRowKeyFields().get(0).getName();
        Object minA = partition1.getRegion().getRange(rowKey).getMin();
        Object minB = partition2.getRegion().getRange(rowKey).getMin();

        if (minA instanceof Long && minB instanceof Long) {
            return ((Long) minA).compareTo((Long) minB);
        }

        if (minA instanceof Integer && minB instanceof Integer) {
            return ((Integer) minA).compareTo((Integer) minB);
        }

        if (minA instanceof String && minB instanceof String) {
            return ((String) minA).compareTo((String) minB);
        }

        if (minA instanceof byte[] && minB instanceof byte[]) {
            return Arrays.compare((byte[]) minA, (byte[]) minB);
        }

        throw new PartitionTreeException("Unable to compare partitions due to range elements being an " +
                "unexpected class type. minA: " + minA.getClass() + ", minB: " + minB.getClass());
    }
}
