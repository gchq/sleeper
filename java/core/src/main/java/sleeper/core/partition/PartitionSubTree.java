/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Objects;

/**
 * Generates a subtree of PartitionTree for given number of leaf nodes.
 */
public class PartitionSubTree extends PartitionTree {
    private final ArrayList<String> resetLeafIds = new ArrayList<>();

    public PartitionSubTree(PartitionTree originalTree, int leafPartitionCount) {
        super(List.of(originalTree.getRootPartition()));

        // Check loop has count incremented by 1 to account for root partition that always must exist
        while (idToPartition.values().size() < leafPartitionCount + 1) {
            resetLeafIds.clear();
            Collection<Partition> presentBatch = List.copyOf(idToPartition.values());
            presentBatch.forEach(partition -> {
                partition.getChildPartitionIds().forEach(
                        partitionId -> {
                            if (!idToPartition.containsKey(partitionId)) {
                                idToPartition.put(partitionId, originalTree.getPartition(partitionId));
                                resetLeafIds.add(partitionId);
                            }
                        });
            });
        }

        // Last gathered selection of ids are set to be leaves of new tree.
        resetLeafIds.forEach(leafId -> {
            idToPartition.put(leafId, idToPartition.get(leafId)
                    .toBuilder()
                    .leafPartition(true)
                    .build());
        });
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        PartitionSubTree that = (PartitionSubTree) o;
        return idToPartition.equals(that.idToPartition);
    }

    @Override
    public int hashCode() {
        return Objects.hash(idToPartition);
    }
}
