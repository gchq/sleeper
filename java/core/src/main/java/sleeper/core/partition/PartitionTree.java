/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import sleeper.core.key.Key;
import sleeper.core.schema.Schema;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.TreeMap;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toUnmodifiableList;

/**
 * Represents a tree of partitions. It can be used to traverse or query the tree, e.g. to find all ancestors of a
 * partition, partitions that are either parents of the partition, or grandparents, or great-grandparents.
 */
public class PartitionTree {
    private final Map<String, Partition> idToPartition;
    private final Partition rootPartition;

    public PartitionTree(Collection<Partition> partitions) {
        this.idToPartition = new TreeMap<>();
        partitions.forEach(p -> this.idToPartition.put(p.getId(), p));
        List<Partition> rootPartitions = partitions.stream().filter(p -> null == p.getParentPartitionId()).collect(Collectors.toList());
        // There should be exactly one root partition.
        if (rootPartitions.size() != 1) {
            throw new IllegalArgumentException("There should be exactly one root partition, found " + rootPartitions.size());
        }
        this.rootPartition = rootPartitions.get(0);
    }

    /**
     * Retrieves IDs of partitions that were split from a given parent partition.
     *
     * @param  partitionId              the ID of the parent partition
     * @return                          the IDs of the child partitions
     * @throws IllegalArgumentException if the parent partition does not exist
     */
    public List<String> getChildIds(String partitionId) throws IllegalArgumentException {
        if (!idToPartition.containsKey(partitionId)) {
            throw new IllegalArgumentException("No partition of id " + partitionId);
        }
        return idToPartition.get(partitionId).getChildPartitionIds();
    }

    /**
     * Retrieves the partitions that were split from a given parent partition.
     *
     * @param  partitionId the ID of the parent partition
     * @return             the child partitions
     */
    public List<Partition> getChildPartitions(String partitionId) {
        List<Partition> outList = new ArrayList<>();
        getChildIds(partitionId).forEach(childId -> {
            outList.add(idToPartition.get(childId));
        });

        return outList;
    }

    /**
     * Retrieves IDs of all partitions that were split to result in the given partition. Starts with the most recent
     * parent, and includes all ancestors of that parent ending with the root partition.
     *
     * @param  partitionId the ID of the partition to find ancestors of
     * @return             all IDs of the partition's ancestors
     */
    public List<String> getAllAncestorIds(String partitionId) {
        if (!idToPartition.containsKey(partitionId)) {
            throw new IllegalArgumentException("No partition of id " + partitionId);
        }
        return ancestorsOf(idToPartition.get(partitionId)).map(Partition::getId).collect(toList());
    }

    /**
     * Retrieves all partitions that were split to result in the given partition. Starts with the most recent parent,
     * and includes all ancestors of that parent ending with the root partition.
     *
     * @param  partitionId the ID of the partition to find ancestors of
     * @return             all of the partition's ancestors
     */
    public List<Partition> getAllAncestors(String partitionId) {
        if (!idToPartition.containsKey(partitionId)) {
            throw new IllegalArgumentException("No partition of id " + partitionId);
        }
        return ancestorsOf(idToPartition.get(partitionId)).collect(toList());
    }

    /**
     * Streams through all partitions that were split to result in the given partition. Starts with the most recent
     * parent, and includes all ancestors of that parent ending with the root partition.
     *
     * @param  partition the ID of the partition to find ancestors of
     * @return           all of the partition's ancestors
     */
    public Stream<Partition> ancestorsOf(Partition partition) {
        return Stream.iterate(getParent(partition),
                currentPartition -> currentPartition != null,
                currentPartition -> getParent(currentPartition));
    }

    /**
     * Streams through all partitions that are descendents of the given partition. Starts with the child partitions,
     * then finds their child partitions, and so on. Traverses depth first.
     *
     * @param  partition the partition
     * @return           the partition's descendents
     */
    public Stream<Partition> descendentsOf(Partition partition) {
        return partition.getChildPartitionIds()
                .stream().map(this::getPartition)
                .flatMap(child -> Stream.concat(Stream.of(child), descendentsOf(child)));
    }

    /**
     * Retrieves the parent of a partition.
     *
     * @param  partition the partition
     * @return           its parent, or null if it has no parent
     */
    public Partition getParent(Partition partition) {
        if (partition.getParentPartitionId() == null) {
            return null;
        }
        return idToPartition.get(partition.getParentPartitionId());
    }

    /**
     * Retrieves a partition by its unique ID.
     *
     * @param  partitionId the ID of the partition
     * @return             the partition
     */
    public Partition getPartition(String partitionId) {
        return idToPartition.get(partitionId);
    }

    public List<Partition> getAllPartitions() {
        return List.copyOf(idToPartition.values());
    }

    public List<Partition> getLeafPartitions() {
        return idToPartition.values().stream().filter(Partition::isLeafPartition).collect(toUnmodifiableList());
    }

    /**
     * Retrieves the leaf partition that includes the given key.
     *
     * @param  schema schema of the Sleeper table
     * @param  key    values for the key to find the leaf partition for
     * @return        the leaf partition
     */
    public Partition getLeafPartition(Schema schema, Key key) {
        // Sanity check key is of the correct length
        if (key.size() != schema.getRowKeyFields().size()) {
            throw new IllegalArgumentException("Key must match the row key fields from the schema (key was "
                    + key + ", schema has row key fields " + schema.getRowKeyFields() + ")");
        }
        // If root partition is a leaf partition then key must be in it
        if (rootPartition.isLeafPartition()) {
            return rootPartition;
        }

        return descend(schema, rootPartition, key);
    }

    private Partition descend(Schema schema, Partition currentNode, Key key) {
        // Get child partitions
        List<String> childPartitionIds = currentNode.getChildPartitionIds();
        List<Partition> childPartitions = new ArrayList<>();
        childPartitionIds.forEach(c -> childPartitions.add(idToPartition.get(c)));

        // Which child is the key in?
        Partition child = null;
        for (Partition partition : childPartitions) {
            if (partition.isRowKeyInPartition(schema, key)) {
                child = partition;
                break;
            }
        }
        if (null == child) {
            throw new IllegalArgumentException("Found key that was not in any of the child partitions: key " + key
                    + ", child partitions " + childPartitions);
        }
        if (child.isLeafPartition()) {
            return child;
        }
        return descend(schema, child, key);
    }

    public Partition getRootPartition() {
        return rootPartition;
    }

    /**
     * Retrieves the partition whose region includes both keys. Finds the nearest partition to each of the keys, with
     * the fewest steps up the partition tree.
     *
     * @param  schema schema of the Sleeper table
     * @param  a      values of the first key
     * @param  b      values of the second key
     * @return        the partition that includes both keys
     */
    public Partition getNearestCommonAncestor(Schema schema, Key a, Key b) {
        return getNearestCommonAncestor(getLeafPartition(schema, a), getLeafPartition(schema, b));
    }

    /**
     * Retrieves the partition whose region includes both descendent partitions. Finds the partition with the fewest
     * steps up the partition tree that is an ancestor for both.
     *
     * @param  a the first partition
     * @param  b the second partition
     * @return   the partition that is an ancestor of both
     */
    public Partition getNearestCommonAncestor(Partition a, Partition b) {
        if (a.getId().equals(b.getId())) {
            return a;
        }
        Set<String> ancestorsB = new HashSet<>(getAllAncestorIds(b.getId()));
        for (String ancestorA : getAllAncestorIds(a.getId())) {
            if (ancestorsB.contains(ancestorA)) {
                return getPartition(ancestorA);
            }
        }
        return getRootPartition();
    }

    /**
     * Traverses the partition tree visiting the leaves first. Proceeds in steps where you remove the current leaf
     * partitions and visit the new leaves.
     * <p>
     * The partitions are also ordered by the min and max of their ranges. Each time the tree is split, the partition
     * on the left/min side of the split will always be displayed first in this ordering.
     * <p>
     * This produces an ordering which is natural to read when you care the most about the leaf partitions, but you also
     * want to be able to read the rest of the tree in a predictable way.
     *
     * @return all partitions in the tree in leaves first order
     */
    public Stream<Partition> traverseLeavesFirst() {
        return traverseLeavesFirst(streamLeavesInTreeOrder().toList(), new HashSet<>(), Stream.empty());
    }

    /**
     * Traverses all leaf partitions by visiting every node in the partition tree. Starts at the root and goes left
     * first then right.
     *
     * @return all leaf partitions in order of their position in the tree
     */
    public Stream<Partition> streamLeavesInTreeOrder() {
        // Establish ordering by combining depth-first tree traversal with the ordering of child IDs on each partition.
        // This should ensure that partitions on the left/min side of a split will always come first in the order.
        return leavesInTreeOrderUnder(getRootPartition());
    }

    private Stream<Partition> leavesInTreeOrderUnder(Partition partition) {
        if (partition.isLeafPartition()) {
            return Stream.of(partition);
        }
        // Always follow left/min side first. Child partition IDs should be left/min then right/max.
        return partition.getChildPartitionIds().stream()
                .map(this::getPartition)
                .flatMap(this::leavesInTreeOrderUnder);
    }

    private Stream<Partition> traverseLeavesFirst(
            List<Partition> leaves, Set<String> prunedIds, Stream<Partition> partialTraversal) {

        // Prune the current leaves from the tree.
        // Tracking the pruned partitions creates a logical tree without needing to update the actual tree.
        leaves.stream().map(Partition::getId).forEach(prunedIds::add);

        // Find the partitions that are the new leaves of the tree after the previous ones were pruned.
        // Ensure the ordering is preserved, as the leaves were given in the correct order.
        List<Partition> nextLeaves = distinctParentsOf(leaves)
                .filter(parent -> prunedIds.containsAll(parent.getChildPartitionIds()))
                .collect(Collectors.toList());

        // Build traversal stream before recursive call, so it's tail-recursive
        Stream<Partition> traversal = Stream.concat(partialTraversal, leaves.stream());
        if (nextLeaves.isEmpty()) {
            return traversal;
        } else {
            return traverseLeavesFirst(nextLeaves, prunedIds, traversal);
        }
    }

    private Stream<Partition> distinctParentsOf(List<Partition> partitions) {
        return partitions.stream()
                .map(Partition::getParentPartitionId).filter(Objects::nonNull)
                .distinct().map(this::getPartition);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        PartitionTree that = (PartitionTree) o;
        return idToPartition.equals(that.idToPartition);
    }

    @Override
    public int hashCode() {
        return Objects.hash(idToPartition);
    }

    @Override
    public String toString() {
        return "PartitionTree{" +
                "idToPartition=" + idToPartition +
                '}';
    }
}
