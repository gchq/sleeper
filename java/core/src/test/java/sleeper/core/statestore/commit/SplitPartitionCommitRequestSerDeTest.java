/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.statestore.commit;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import sleeper.core.partition.PartitionTree;
import sleeper.core.partition.PartitionsBuilder;
import sleeper.core.schema.Schema;
import sleeper.core.schema.type.StringType;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static sleeper.core.schema.SchemaTestHelper.schemaWithKey;

public class SplitPartitionCommitRequestSerDeTest {

    private final Schema schema = schemaWithKey("key", new StringType());
    private final SplitPartitionCommitRequestSerDe serDe = new SplitPartitionCommitRequestSerDe(schema);

    @Test
    void shouldSerialiseSplitPartitionRequest() {
        // Given
        PartitionTree partitionTree = new PartitionsBuilder(schema)
                .rootFirst("root")
                .splitToNewChildren("root", "left", "right", "aaa")
                .buildTree();
        SplitPartitionCommitRequest splitPartitionCommitRequest = new SplitPartitionCommitRequest(
                "test-table",
                partitionTree.getRootPartition(),
                partitionTree.getPartition("left"), partitionTree.getPartition("right"));

        // When
        String json = serDe.toJsonPrettyPrint(splitPartitionCommitRequest);

        // Then
        assertThat(serDe.fromJson(json)).isEqualTo(splitPartitionCommitRequest);
        Approvals.verify(json);
    }

    @Test
    void shouldFailToDeserialiseNonSplitPartitionCommitRequest() {
        assertThatThrownBy(() -> serDe.fromJson("{\"type\": \"OTHER\", \"request\":{}}"))
                .isInstanceOf(IllegalArgumentException.class);
    }
}
