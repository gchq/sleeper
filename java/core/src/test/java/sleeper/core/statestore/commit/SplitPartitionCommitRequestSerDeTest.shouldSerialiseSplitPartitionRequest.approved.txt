{
  "type": "SPLIT_PARTITION",
  "request": {
    "tableId": "test-table",
    "parentPartition": {
      "partitionId": "root",
      "isLeafPartition": false,
      "parentPartitionId": null,
      "childPartitionIds": [
        "left",
        "right"
      ],
      "region": {
        "key": {
          "min": "",
          "minInclusive": true,
          "max": null,
          "maxInclusive": false
        },
        "stringsBase64Encoded": true
      },
      "dimension": 0
    },
    "leftChild": {
      "partitionId": "left",
      "isLeafPartition": true,
      "parentPartitionId": "root",
      "childPartitionIds": [],
      "region": {
        "key": {
          "min": "",
          "minInclusive": true,
          "max": "YWFh",
          "maxInclusive": false
        },
        "stringsBase64Encoded": true
      },
      "dimension": -1
    },
    "rightChild": {
      "partitionId": "right",
      "isLeafPartition": true,
      "parentPartitionId": "root",
      "childPartitionIds": [],
      "region": {
        "key": {
          "min": "YWFh",
          "minInclusive": true,
          "max": null,
          "maxInclusive": false
        },
        "stringsBase64Encoded": true
      },
      "dimension": -1
    }
  }
}